File: src/2359.c
#include <limits.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void dfs(int node, int* edges, int* dist, int* visit) {
    visit[node] = 1;

    int neighbor = edges[node];
    if ((neighbor != -1) && (visit[neighbor] == 0)) {
        dist[neighbor] = 1 + dist[node];
        dfs(neighbor, edges, dist, visit);
    }
}
int closestMeetingNode(int* edges, int edgesSize, int node1, int node2) {
    int retVal = -1;

    int i;

    int dist1[edgesSize];
    for (i = 0; i < edgesSize; ++i) {
        dist1[i] = INT_MAX;
    }
    dist1[node1] = 0;
    int visit1[edgesSize];
    memset(visit1, 0, sizeof(visit1));
    dfs(node1, edges, dist1, visit1);

    int dist2[edgesSize];
    for (i = 0; i < edgesSize; ++i) {
        dist2[i] = INT_MAX;
    }
    dist2[node2] = 0;
    int visit2[edgesSize];
    memset(visit2, 0, sizeof(visit1));
    dfs(node2, edges, dist2, visit2);

    int minDist = INT_MAX;
    for (i = 0; i < edgesSize; ++i) {
        if (minDist > fmax(dist1[i], dist2[i])) {
            retVal = i;
            minDist = fmax(dist1[i], dist2[i]);
        }
    }

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e5)
    struct testCaseType {
        int edges[MAX_SIZE];
        int edgesSize;
        int node1;
        int node2;
    } testCase[] = {{{2, 2, 3, -1}, 4, 0, 1}, {{1, 2, -1}, 3, 0, 2}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);

    int answer = 0;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: edges =  [");
        for (j = 0; j < testCase[i].edgesSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].edges[j]);
        }
        printf("], node1 = %d, node2 = %d\n", testCase[i].node1, testCase[i].node2);

        answer = closestMeetingNode(testCase[i].edges, testCase[i].edgesSize, testCase[i].node1, testCase[i].node2);
        printf("Output: %d\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/2359.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def dfs(self, node: int, edges: List[int], dist: List[int], visit: List[bool]) -> None:
        visit[node] = True

        neighbor = edges[node]
        if (neighbor != -1) and (visit[neighbor] == False):
            dist[neighbor] = 1 + dist[node]
            self.dfs(neighbor, edges, dist, visit)

    def closestMeetingNode(self, edges: List[int], node1: int, node2: int) -> int:
        retVal = -1

        edgesSize = len(edges)

        dist1 = [float('inf')] * edgesSize
        dist1[node1] = 0
        visit1 = [False] * edgesSize
        self.dfs(node1, edges, dist1, visit1)

        dist2 = [float('inf')] * edgesSize
        dist2[node2] = 0
        visit2 = [False] * edgesSize
        self.dfs(node2, edges, dist2, visit2)

        minDist = float('inf')
        for currNode in range(edgesSize):
            if minDist > max(dist1[currNode], dist2[currNode]):
                retVal = currNode
                minDist = max(dist1[currNode], dist2[currNode])

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for edges, node1, node2 in zip([[2, 2, 3, -1], [1, 2, -1]], [0, 0], [1, 2]):
            # /* Example
            #  *  Input: edges = [2,2,3,-1], node1 = 0, node2 = 1
            #  *  Output: 2
            #  *
            #  *  Input: edges = [1,2,-1], node1 = 0, node2 = 2
            #  *  Output: 2
            #  */
            logging.info("Input: nums = %s, node1 = %s, node2 = %s", edges, node1, node2)

            retVal = pSolution.closestMeetingNode(edges, node1, node2)
            logging.info("Output: %d", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
